README
MString - Dynamic string data type library
Copyright (C) 2001-2005 Jesse L. Lovelace 
                        (thementat at users.sourceforge.net)

http://mstring.sf.net

MString is a sizeable string library implimented via linked lists. For a
complete list of functions please see doc/FUNCTIONS.MString . MString is
released under the terms of the LGPL (license).

MStringArray is a dynamic array of MString objects.

MString is currently under development!  Please submit any
bugs/suggestions/comments to:

 Jesse [thementat at users.sourceforge.net] 
   or
 Brian Matherly [pez4brian at users.sourceforge.net]

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


NEWS
April 5, 2005 - Moving to the LGPL

Dec. 1, 2004 - Brian Matherly has joined the project as
     a project manager and updated the code with lots of
     fixes and such.

Sept. 4, 2000 - After a summer break I'm now ready to
     start developing MString full speed.  Thanks to
     everyone who wrote bug reports!


test/driver.cpp
/*  driver.cpp - Run time tests for MString
    Copyright (C) 2001-2005 Jesse L. Lovelace (jesse at aslogicsys dot com)

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include "../src/MString.h"
#include "../src/MStringArray.h"
#include <fstream.h>

void main() 
{

	//This file tests *all* MString functions (wip!)

	char myChar[] = "Hello, this is 1234, a string...?";
	char myLet = 'j';

	MString fromChar(myChar);
	MString fromLet(myLet);
	MString fromLetRep(myLet,20);
	MString fromM(fromChar);

	cout << fromChar << endl
		<< fromLet << endl
		<< fromLetRep << endl
		<< fromM << endl;

	//Large file read and memory test.
/*


	MString readLarge;

	for (int f1 = 0; f1 < 1000; f1++) {
		ifstream in;
		in.open("h:\\server\\MString\\largefile.txt");
		while (in) {
		
			in >> readLarge;
		}
		cout << readLarge;
		readLarge.Empty();
		
		cout << "!!!!    " << f1 << endl;
	}

*/
	
	//Loop test for deallocation
	//also Empty()
	MString test;
	for (int i = 0; i < 3000; i++) {
		test = "this is a test";
		test.Empty();
	}

	//GetLength
	cout << "Length of fromChar = " << fromChar.GetLength()
		<< endl;

	//IsEmpty
	cout << "IsEmpty()\n";
	if (fromChar.IsEmpty())
		cout << "!!!! fromChar is EMPTY\n";
	else cout << "fromChar is not EMPTY\n";

	//GetAt
	cout << "GetAt\n";
	for (int k = 0; k < fromChar.GetLength(); k++) {
		cout << fromChar.GetAt(k);
	}
	cout << endl;
	
	//Rvalue operator []

	cout << "Rvalue operator []\n";

	for (int f = 0; f < fromLetRep.GetLength(); f++) {
		fromLetRep[f] = 'w';
	}
	cout << fromLetRep << endl;


	//SetAt

	cout << "SetAt\n";

	for (int g = 0; g < fromLetRep.GetLength(); g++) {
		fromLetRep.SetAt(g, 'j');
	}
	cout << fromLetRep << endl;

	// = 

	cout << " = \n";

	test = fromLetRep;
	cout << test << endl;
	test = 'y';
	cout << test << endl;
	test = "hello and how are you?";
	cout << test << endl;
	test = 675444;
	cout << test << endl;
	test = 323.2341;
	cout << test << endl;

	// + 

	cout << " + \n";

	test = fromLetRep + fromChar;
	cout << test << endl;
	test = fromChar + '>';
	cout << test << endl;
	test = '<' + fromChar;
	cout << test << endl;
	test = fromChar + "CHAR STRING";
	cout << test << endl;
	test = "CHAR STRING" + fromChar;
	cout << test << endl;

	// +=

	cout << " += \n";

	test += fromLetRep;
	cout << test << endl;
	test.Empty();
	test += 'j';
	cout << test << endl;
	test.Empty();
	test += "CHAR STRING????";
	cout << test << endl;
	test += 999999;
	cout << test << endl;
	test.Empty();
	test += 34.2342;
	cout << test << endl;	

	test.Empty();
	test = 34.00000234;
	cout << test << endl;
	test.Empty();
	test = 340.012;
	cout << test << endl;

	MString dub(456.322);
	cout << dub << endl;	
	dub.Replace('2','6');
	cout << dub << endl;
}


src/MString.h
/*  MString - Dynamic string data type library
    Copyright (C) 2001-2005 Jesse L. Lovelace

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifndef MSTRING_H
#define MSTRING_H

#ifdef __cplusplus  // we only want this for C++ code

/*  MString by Jesse Lovelace (mentat)
	jesse at aslogicsys dot com

    (MString began as GString but was changed to remain compatible with
	gtk+'s GString data type.)

	http://mstring.sourceforge.net

	MString is a string library that uses dynamically sizable character
	strings.  To do this I used linked lists instead of dynamic arrays for
	various reasons (one being the problems that arrays of dynamic arrays
	cause).

	MString is designed to maintain compatibility with the MFC CString
	library.  MString should be 99% compatible if you exclude the windoze
	specific operations (including buffer access) and the .Format commands.

	For the full documentation see functions.txt

	MString also includes the MStringArray class.  MStringArray follows the
	same scheme as CStringArray from MFC.  CStringArrays are also linked
	lists for now (I may change them to dynamically allocated arrays in the
	future).

	Please submit any suggestions/bugs/comments to jllovela@eos.ncsu.edu

	Enjoy.

	Thanks to:
      Brian Matherly - Project manager and developer.  
      Balint Toth - Fixed insert bug and new features,
      Bruce Riggins - New char * code,
      Dingo - pointer debugging, 
      Rask - math and Trim(char ch), 
      Botch - debugging and advice,
      Antiloop - for phat grooves.

	(Last modified April 10, 2005)
*/

const int MAX_PRECISION = 9; //due to bad double to string code.
const char MString_VERSION[8] = "0.70";
//typedef const char * LPCSTR;

#include <fstream>

class MNode;

struct MStringData {
	char * pcStr;
	int precision;
	int iBufLen; // length of buffer minus 1 for null terminator
	int iBufferInUse;
};

class MString {

//Begin overloaded stream functions -----------------------

	friend std::ostream& operator<< (std::ostream& out, const MString& string); 
	// PRE: out is a valid output stream, string is a valid MString
	// POST: contents of string are sent to stream out.

	friend std::istream& operator>> (std::istream& in, MString& string); 
	// PRE: in is a valid input stream, string is a valid MString
	// POST: characters in input stream are stored in string.

	friend std::istream& getline(std::istream& in, MString& string); 
	// PRE: in is valid input stream, sting is a valid MString
	// POST: characters up to (and including) the newline are stored 
	//       in string.
	// EX: getline(in, string);

//End overloaded stream functions -------------------------

public:

//Begin Constructors --------------------------------------

	MString(); 
	// Info: construct empty MString
	
	MString(const MString& stringSrc); 
	// Info: copy constructor
	// PRE: stringSrc is a valid MString object.
	// POST: new object is a deep copy of stringSrc.

	MString(const char ch, int nRepeat = 1);
	// Info: construct with single character repeated nRepeat times
	// PRE: ch is a valid ascii character and non null.
	// POST: new object contains ch repeated nRepeat times.

	MString(const char* string);
	// PRE: string is a valid null-terminated string.
	// POST: new object contains the data from string minues null.

// Added by Bruce Riggins	
	MString(const char * string, int nLength);
	// Info: copy with maximum length
	// PRE: string is a valid null-terminated string, nLength>0
	// POST: new object contains the data from string minues null
// End addition	
	
	MString(const int num);
	// NEW CODE, PLEASE HELP DEBUG (v. 60)
	MString(double num);

	~MString(); 
	// Info: destructor

//End Constructors ----------------------------------------

//Begin Test Functions ------------------------------------

	void testG();
	// Info: uses cout to display the data and pointer of each
	//       Node.

	MString GetGVersion();
	// Info: returns the current version of MString.
	// POST: returns an object of type MString containing version info

	MString ReturnLicense();
	// Info: returns license info
	// POST: returns an object of type MString containing license info


//End Test Functions --------------------------------------

//Begin Precision Functions -------------------------------

	void SetPrecision(int num);
	// PRE: num <= MAX_PRECISION
	// POST: double conversion precision changed to num.

	int GetPrecision() const;
	// POST: returns an int type equal to the current double precision.

//End Precision Functions ---------------------------------

//Begin String as Array Functions -------------------------
	
	int GetLength() const; 
	// Info: returns the length of the string excluding null

	bool IsEmpty() const;  
	// Info: nonzero if string has 0 length
	
	void Empty();  
	// Info: empty's the string and free's memory
	
	char GetAt(int nIndex) const; 
	// Info: returns char at index
	
	char operator [](int nIndex) const; 
	// Info: operator overload of GetAt(int nIndex)

	char& operator [](int nIndex);
	// Info: use like SetAt, myString[0] = 'M';

	void SetAt(int nIndex, char ch); 
	// Info: sets the character at nIndex

//End String as Array Functions ---------------------------

//Begin added operators by Bruce Riggins ---------------
	operator const char * () const;
//End added operators by Bruce Riggins -----------------

//Begin Assignment/Concatination operators ----------------

	const MString& operator =(const MString& stringSrc); //Idea from CString
	const MString& operator =(char ch);
	const MString& operator =(const char* string);
	const MString& operator =(int num); //Original MString
	const MString& operator =(double num); //Original MString
	const MString& operator =(float num); //Original MString

	const MString& operator +=(const MString& string); //Idea from CString
	const MString& operator +=(char ch);
	const MString& operator +=(char* string);
	const MString& operator +=(int num); //Original MString
	const MString& operator +=(double num); //Original MString
	const MString& operator +=(float num); //Original MString

	void Cat(char ch);   //Not implimented yet.
	void Cat(char* string);
	void Cat(MString& string); 

	

	friend MString operator +(const MString& string1, const MString& string2);
	friend MString operator +(const MString& string, char ch);
	friend MString operator +(char ch, const MString& string);
	friend MString operator +(const MString& string1, char* string2);
	friend MString operator +(char* string1, const MString& string2);
	friend MString operator +(int num, const MString& string);
	friend MString operator +(const MString& string, int num);


//End Assignment/Concatination operators ------------------

//Begin Comparison operators ------------------------------

	friend bool operator==(const MString& s1, const MString& s2); //Idea from CString
	friend bool operator==(const MString& s1, const char* s2);
	friend bool operator==(const char* s1, const MString& s2);

	friend bool operator!=(const MString& s1, const MString& s2); //Idea from CString
	friend bool operator!=(const MString& s1, char* s2);
	friend bool operator!=(char* s1, const MString& s2);
	friend bool operator!=(const MString& s1, const char* s2);

	friend bool operator <(const MString& s1, const MString& s2); //Idea from CString
	friend bool operator <(const MString& s1, char* s2);
	friend bool operator <(char* s1, const MString& s2);

	friend bool operator >(const MString& s1, const MString& s2); //Idea from CString
	friend bool operator >(const MString& s1, char* s2);
	friend bool operator >(char* s1, const MString& s2);

	friend bool operator <=(const MString& s1, const MString& s2); //Idea from CString
	friend bool operator <=(const MString& s1, char* s2);
	friend bool operator <=(char* s1, const MString& s2);

	friend bool operator >=(const MString& s1, const MString& s2); //Idea from CString
	friend bool operator >=(const MString& s1, char* s2);
	friend bool operator >=(char* s1, const MString& s2);

	int Compare(const char* string) const; //Idea from CString
	int Compare(const MString& string) const; //MString original
	int CompareNoCase(const char* string) const; //Idea from CString
	int Collate(char* string) const; //Idea from CString
	int CollateNoCase(char* string) const; //Idea from CString

	bool CompareGlob(const MString& glob) const;

//End Comparison operators --------------------------------

//Begin Extraction operators ------------------------------

	MString Mid(int nFirst) const;  //Idea from CString
	MString Mid(int nFirst, int nCount) const; //Idea from CString
	MString Left(int nCount) const; //Idea from CString
	MString Right(int nCount) const; //Idea from CString
	MString SpanIncluding(char* string) const; //Idea from CString
	MString SpanExcluding(char* string) const; //Idea from CString
	char* ToChar(int nStart = 0,int nCount = -1); 
	// PRE: nStart is >0 and < GetLength().
	// POST: returns a pointer to a new null-terminated character array
	//       that is a copy of the infomation in the MString.

	int ToInt(int nStart = 0);
	// PRE: nStart is > 0 and < GetLength().  
	// POST: returns an integer created from the numbers starting at nStart
	// Other: ToInt does not check to see if the string of numbers contained 
	//        in the MString object is to long for type int.  ToInt converts
	//        a sequence of ASCII values 48 thru 57 ('0' to '9') non-number
	//        values result in a return from the function.
	// EX: if myString is of type MString containing "12345C1232"
	//     myString.ToInt(0) returns 12345
	//     myString.ToInt(1) returns 2345

	double ToDouble( int nStart = 0 );

//End Extraction operators --------------------------------

//Begin Other Conversions ---------------------------------
	
	void MakeUpper(); //Idea from CString
	void MakeLower(); //Idea from CString
	void MakeReverse(); //Idea from CString

	int Replace(char chOld, char chNew); //Idea from CString
	int Replace(char* stringOld, char* stringNew); //Idea from CString
	int Remove(char ch); //Idea from CString
	int Insert(int nIndex, char ch); //Idea from CString
	int Insert(int nIndex, char* string); //Idea from CString
	int Delete(int nIndex, int nCount = 1); //Idea from CString

	//Research Format
// Additions by Bruce Riggins 11/14/00	
	void Format(const char * sFormat, ...);
// End additions by Bruce Riggins 11/14/00	

	void Trim(); //Original MString
	void Trim(char ch); //Original MString
	void Trim(char* string);

	void TrimLeft(); //Idea from CString
	void TrimLeft(char ch);
	void TrimLeft(char* string);


	void TrimRight(); //Idea from CString
	void TrimRight(char ch);
	void TrimRight(char* string);

//End Other Conversions -----------------------------------

	//Research FormatMessage

//Begin Searching -----------------------------------------

	int Find(char ch, int nStart = 0) const; //Idea from CString
	int Find(char* string, int nStart = 0) const; //Idea from CString


	int ReverseFind(char ch) const; //Idea from CString
	int ReverseFind(char* string) const; //MString original
	int FindOneOf(char* string) const; //Idea from CStrin


//End Searching -------------------------------------------

	//Buffer Access and Windows-Specific items not included.

//Begin Buffer Access -------------------------------------
	// Additions by Bruce Riggins 11/7/00
	// Access to string implementation buffer as "C" character array
	char * GetBuffer(int nMinBufLength = -1);	//Idea from CString
	void ReleaseBuffer(int nNewLength = -1);	//Idea from CString
//	char * GetBufferSetLength(int nNewLength);	//Idea from CString
//	void FreeExtra();	//Idea from CString

	// Use LockBuffer/UnlockBuffer to turn refcounting off
	char * LockBuffer();	//Idea from CString
	void UnlockBuffer();	//Idea from CString
	
	// End of BR additions

//End Buffer Access ---------------------------------------

private:

	MNode* GetPointerAt(int nIndex);
	void deallocate(MNode *p); // added by Bruce Riggins

	bool Gmatch( MNode* g, MNode* s ) const;
	
	char * pcStr;	
	MNode *headMNode;
	MNode *tailMNode; //New for .3b

	int precision;
	int iBufLen; // length of buffer minus 1 for null terminator
	int iBufferInUse;

   // flag to indicate that the MString has been modified since the
   // last time char* GetBuffer() was called
   bool bModified; 

};

#endif // __cplusplus

#endif


src/MStringArray.h
/*  MStringArray - Dynamic Array of MString objects
    Copyright (C) 2001-2005 Jesse L. Lovelace

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifndef MSTRINGARRAY_H
#define MSTRINGARRAY_H

#include "MString.h"

const char MStringArray_VERSION[8] = "BETA 1";

class MArrayNode;


class MStringArray {
public:
	MStringArray();
	MStringArray(const MStringArray& arraySrc);
	MStringArray(const MString& stringSrc);
	~MStringArray();

	const MStringArray& operator =(const MStringArray& arraySrc);
	int GetSize() const;
	int GetUpperBound() const;
	void SetSize(int nNewSize, int nGrowBy = -1);

	void FreeExtra();
	void RemoveAll();
	MString GetAt(int nIndex) const; 
	void SetAt(int nIndex, const MString &newElement);
	void SetAt(int nIndex, char newString[]);
	MString& ElementAt(int nIndex);

//	const MString* GetData() const;
//	MString* GetData();


	void SetAtGrow(int nIndex, const MString &newElement);
	void SetAtGrow(int nIndex, char newString[]);
	int Add(const MString &newElement);
	int Add(char newString[]);
	int Append(const MStringArray &src);
	void Copy(const MStringArray &src);

	void InsertAt(int nIndex, const MString &newElement, int nCount = 1);
	void InsertAt(int nStartIndex, MStringArray pNewArray);
	void InsertAt(int nIndex, char newString[], int nCount = 1);

	void RemoveAt(int nIndex, int nCount = 1);

	MString& operator[](int nIndex);
	MString operator[](int nIndex) const;

	//New functions:

	void Split(MString stringToSplit, char ch = ' ',  int nStartIndex = 0, int insertAt = 0);

	void test();

private:
	MArrayNode* GetPointerAt(int nIndex);

	MArrayNode *headMNode;
	MArrayNode *tailMNode;

};

#endif


src/MStringArray.cpp
/*  MStringArray - Dynamic Array of MString objects
    Copyright (C) 2001-2005 Jesse L. Lovelace (jesse at aslogicsys dot com)

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <fstream>
#include <iostream>
#include "MStringArray.h"
#include "MString.h"

using namespace std;

class MArrayNode {
public:
	MArrayNode(MString string, MArrayNode* link = 0) {MLink_Forward = link; MInfo = string;}
	MArrayNode *MLink_Forward;
	MString MInfo;
};

static void deallocate(MArrayNode* p)
{
	MArrayNode* tmp;
	while (p) {
		tmp = p;
		p = p->MLink_Forward;
		delete tmp;
	}
}

static MArrayNode* findTail(MArrayNode * startPointer) {
	MArrayNode * tmp=startPointer;
	while (tmp) {
		if (!tmp->MLink_Forward)
			return tmp;
		tmp = tmp->MLink_Forward;
	}
	return startPointer;
}

static MArrayNode* copy(MArrayNode* p, MArrayNode*& tailNode)
{
	MArrayNode* head = 0;
	MArrayNode* tail = 0;


	while(p) {
		if (!head)
			head = tail = new MArrayNode(p->MInfo);
		else {
			tail->MLink_Forward = new MArrayNode(p->MInfo);
			tail = tail->MLink_Forward;
		}
		p = p->MLink_Forward;
	}
	tailNode = tail;
	return head;
}

MArrayNode* MStringArray::GetPointerAt(int nIndex) {

	MArrayNode* tmp = headMNode;
	
	for (int i = 0; (i < nIndex); i++) {

		if (NULL == tmp)
			return NULL;

		tmp = tmp->MLink_Forward;
	}
	return tmp;
}


MStringArray::~MStringArray() {
	deallocate(headMNode);
	headMNode = tailMNode = NULL;
}

MStringArray::MStringArray() {

	headMNode = tailMNode = NULL;
}

MStringArray::MStringArray(const MStringArray& arraySrc) {

	headMNode = copy(arraySrc.headMNode, tailMNode);
}

MStringArray::MStringArray(const MString& stringSrc) {

	headMNode = tailMNode = new MArrayNode(stringSrc);
}

const MStringArray& MStringArray::operator = (const MStringArray& arraySrc) {

	if (this != &arraySrc) {
		deallocate(headMNode);
		headMNode = tailMNode = NULL;
		headMNode = copy(arraySrc.headMNode, tailMNode);
	}
	return *this;
}


int MStringArray::GetSize() const {

	int count = 0;

	if (headMNode==NULL)
		return count;

	MArrayNode *tmp = headMNode;

	while(tmp) {
		count++;
		tmp = tmp->MLink_Forward;
	}
	return count;

}

int MStringArray::GetUpperBound() const {

	return (GetSize() - 1);
}

void MStringArray::SetSize(int nNewSize, int nGrowBy) {

	if (nNewSize < 0)
		return;
	//UNFINISHED FUNCTION
	//nGrowBy unreferenced formal param

}


void MStringArray::FreeExtra() {

	//UNFINISHED FUNCTION

}

void MStringArray::RemoveAll() {

	deallocate(headMNode);
	headMNode = tailMNode = NULL;

}

void MStringArray::test() {

	MArrayNode *tmp = headMNode;

	int count = 0;
	cout << "Head = " << headMNode << endl;
	while (tmp)
	{ 
		cout << "Index(" << count++ << ") " << tmp->MInfo << " " << tmp->MLink_Forward << endl;
		tmp = tmp->MLink_Forward;
	}
	cout << "Tail = " << tailMNode << endl;
	cout << "GetSize = " << GetSize() << endl;
}
	


MString MStringArray::GetAt(int nIndex) const {
	
	MString tmpStr;
	if ((nIndex < 0) || (nIndex > GetUpperBound()))
		return tmpStr;   //return's null MString if the nIndex is to big or small.

	MArrayNode* tmp = headMNode;

	for (int i = 0; i < nIndex; i++)
		tmp=tmp->MLink_Forward;

	return tmp->MInfo;
}

void MStringArray::SetAt(int nIndex, const MString &newElement) {

	if ((nIndex < 0) || (nIndex > GetUpperBound()))
		return;   //return's null MString if the nIndex is to big or small.

	MArrayNode* tmp = headMNode;

	for (int i = 0; i < nIndex; i++)
		tmp=tmp->MLink_Forward;

	tmp->MInfo = newElement;
}

void MStringArray::SetAt(int nIndex, char newString[]) {

	if ((nIndex < 0) || (nIndex > GetUpperBound()))
		return;   //return's null MString if the nIndex is to big or small.

	MArrayNode* tmp = headMNode;
	
	for (int i = 0; i < nIndex; i++)
		tmp=tmp->MLink_Forward;

	tmp->MInfo = newString;
}


MString& MStringArray::ElementAt(int nIndex) {
	
	MString* tmpStr = NULL;
	if ((nIndex < 0) || (nIndex > GetUpperBound()))
		return *tmpStr;   //return's null MString if the nIndex is to big or small.

	MArrayNode* tmp = headMNode;

	for (int i = 0; i < nIndex; i++)
		tmp=tmp->MLink_Forward;

	return tmp->MInfo;
}

// const MString* MStringArray::GetData() const;
// MString* MStringArray::GetData();


void MStringArray::SetAtGrow(int nIndex, const MString &newElement) {

	//Not written yet
}

void MStringArray::SetAtGrow(int nIndex, char newString[]) {
	//Not written yet

}

int MStringArray::Add(const MString &newElement) {

	InsertAt(GetSize(), newElement);

	return GetUpperBound();
}

int MStringArray::Add(char newString[]) { //works

	InsertAt(GetSize(), newString);

	return GetUpperBound();
}


int MStringArray::Append(const MStringArray &src) { //works

	InsertAt(GetSize(), src);

	return GetSize();
}

void MStringArray::Copy(const MStringArray &src) {

	headMNode = copy(src.headMNode, tailMNode);

}
void MStringArray::InsertAt(int nIndex, const MString &newElement, int nCount) { //works
	//MFC Uses pointer to MString, 
	
	int length = GetSize();

	if ((nCount == 0) || (nIndex < 0))
		return;

	if (length == 0) {  //if size is zero
		headMNode = tailMNode = new MArrayNode(newElement);
		return;
	}


	if (nIndex == length) {//if just adding to end, use tailMNode
		for (int a = 0; a < nCount; a++) {
			tailMNode->MLink_Forward = new MArrayNode(newElement);
			tailMNode = tailMNode->MLink_Forward;
		}
		return;
	}

	if (nIndex == 0) {
		headMNode = new MArrayNode(newElement, headMNode);
		MArrayNode* tmp = headMNode;

		for (int i = 1; i < nCount; i++) {
			tmp->MLink_Forward = new MArrayNode(newElement, tmp->MLink_Forward);
			tmp = tmp->MLink_Forward;
		}
		return;

	}

	
	if (nIndex > (length - 1)) {//if space between new element and old ones,
								 // pad with null MStrings
		
		for (int adds = nIndex; nIndex > GetSize(); adds++) {
			MString tmpStr;
			Add(tmpStr);
		}
	
		for (int other = 0; other < nCount; other++) {
			Add(newElement);
		}
		return;
	}

	// if inserting in the middle somewhere
	MArrayNode * tmp = headMNode;
	for (int k = 0; k < (nIndex - 1); k++) { //go to index you want to insert infront of.
		tmp = tmp->MLink_Forward;
	}
	for (int w = 0; w < nCount; w++) {
		tmp->MLink_Forward = new MArrayNode(newElement, tmp->MLink_Forward);
		tmp = tmp->MLink_Forward;
	}
	return;

}
		
			
void MStringArray::InsertAt(int nStartIndex, MStringArray pNewArray){ //works

	int size = pNewArray.GetSize();

	for (int i = 0; i < size; i++)
		InsertAt(nStartIndex++, pNewArray.GetAt(i)); //sloppy jesse, don't use GetAt
}


void MStringArray::InsertAt(int nIndex, char newString[], int nCount) {
	

	MString tmp(newString);
	InsertAt(nIndex, tmp, nCount);

	return;

}

void MStringArray::RemoveAt(int nIndex, int nCount) {

	int length = GetSize();

	if ((length < 1) || (nIndex > length) || (nCount < 1))
		return;

	if (nCount > (length - nIndex))
		nCount = (length - nIndex);

	if (0 == nIndex) {
		MArrayNode* first = headMNode;
		MArrayNode* last = GetPointerAt((nIndex + nCount)-1);
	
		if (last->MLink_Forward) {
			headMNode = last->MLink_Forward;
			last->MLink_Forward = NULL;
			deallocate(first);
		}
		else {
			headMNode = NULL;
			deallocate(first);
		}

	}	
	else {
		MArrayNode* toHead = GetPointerAt(nIndex - 1);
		MArrayNode* first = GetPointerAt(nIndex);

		MArrayNode* last = GetPointerAt((nCount + nIndex) - 1);
	
		if (last->MLink_Forward) {
			if (last->MLink_Forward->MLink_Forward) {
				toHead->MLink_Forward = last->MLink_Forward;
				last->MLink_Forward = NULL;
				deallocate(first);
			}
			else {
				toHead->MLink_Forward=last->MLink_Forward;
				last->MLink_Forward = NULL;
				deallocate(first);
			}
		}
		else {
			toHead->MLink_Forward = NULL;
			deallocate(first);
		}
	}
	tailMNode = findTail(headMNode);

}

MString& MStringArray::operator[](int nIndex) {
	
	MString* tmpStr = NULL;
	if ((nIndex < 0) || (nIndex > GetUpperBound()))
		return *tmpStr;   //return's null MString if the nIndex is to big or small.

	MArrayNode* tmp = headMNode;

	for (int i = 0; i < nIndex; i++)
		tmp=tmp->MLink_Forward;

	return tmp->MInfo;
}

MString MStringArray::operator[](int nIndex) const {
	
	MString tmpStr;
	if ((nIndex < 0) || (nIndex > GetUpperBound()))
		return tmpStr;   //return's null MString if the nIndex is to big or small.

	MArrayNode* tmp = headMNode;

	for (int i = 0; i < nIndex; i++)
		tmp=tmp->MLink_Forward;

	return tmp->MInfo;
}

	//New functions:

void MStringArray::Split(MString stringToSplit, char ch, int nStartIndex, int insertAt){

	MStringArray tmpArray;

	MString tmpString;

	int place = nStartIndex,
		nextFind = stringToSplit.Find(ch, place);

	while (nextFind >= 0) {

		tmpString = stringToSplit.Mid(place, nextFind - place);

		tmpString.Trim(ch);

		cout << tmpString << endl;
		cout << "place = " << place << endl;

		tmpArray.Add(tmpString);

		tmpString.Empty();

		place = nextFind + 1;
		nextFind = stringToSplit.Find(ch, place);

		cout << "place = " << place << endl;
	}
	tmpString = stringToSplit.Right(stringToSplit.GetLength() - place); 
	//get the last word and add
	// it
	tmpString.Trim(ch);
	if (!tmpString.IsEmpty()) {
		tmpArray.Add(tmpString);
		tmpString.Empty();
	}

	InsertAt(insertAt, tmpArray);
}
